diff --git a/deepspeed/utils/numa.py b/deepspeed/utils/numa.py
index 13617826..e89a40f6 100644
--- a/deepspeed/utils/numa.py
+++ b/deepspeed/utils/numa.py
@@ -165,7 +165,7 @@ def get_numactl_cmd(bind_core_list, num_local_procs, local_rank):
     elif numa_mode == "flat_hbm":
         for i in range(num_numas):
             if set(core_list_for_rank) <= set(numa_cores[i]):
-                numactl_cmd.append("-p")
+                numactl_cmd.append("-m")
                 numactl_cmd.append(f"{numa_dict[i]}")
                 break
     elif numa_mode == "fake":
